User installed fonts are not always registered

pvollan opened the pull request:
#### 1a1486ea60b28c3f4d64024f4635adbe3d2fe9b3
<pre>
User installed fonts are not always registered
<a href="https://bugs.webkit.org/show_bug.cgi?id=286511">https://bugs.webkit.org/show_bug.cgi?id=286511</a>
<a href="https://rdar.apple.com/143598462">rdar://143598462</a>

Reviewed by Alexey Proskuryakov.

User installed fonts are not always registered because we are doing a case sensitive comparison.

* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::registerUserInstalledFonts):

Canonical link: <a href="https://commits.webkit.org/289691@main">https://commits.webkit.org/289691@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/710e1a902e82728bdd2b76909671b8778f4e6a9c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87620 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7135 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42001 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92485 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38364 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7516 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15305 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67663 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25404 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90622 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5734 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79289 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48026 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5522 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37477 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75922 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34551 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94371 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14788 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10841 "Found 5 new test failures: fast/dom/crash-with-bad-url.html fast/text/font-face-set-cssom.html ipc/create-media-source-with-invalid-constraints-crash.html media/media-session/actionHandler-no-document-leak.html svg/custom/use-image-in-g.svg (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76511 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15042 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75142 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75734 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20107 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18541 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7739 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13677 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14804 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/20105 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14548 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17992 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16330 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1381,9 +1381,12 @@ static AdditionalFonts additionalFonts(const Vector<URL>& fontURLs, std::optiona
+            RetainPtr fontsPath = [path stringByAppendingPathComponent:font];
+            auto fontURL = URL::fileURLWithFileSystemPath(String(fontsPath.get()));
+            NSURL *nsFontURL = fontURL;
+            if ([nsFontURL.pathExtension caseInsensitiveCompare:@"ttf"] == NSOrderedSame
```
Is there a way to delegate extension checking to underlying frameworks, so that we'd be looking at UTIs?

pvollan replied:
That is a very good point! I will look closer at that.

Thanks for reviewing!

aproskuryakov commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1429,9 +1430,11 @@ static void addUserInstalledFontURLs(NSString *path, Vector<URL>& fontURLs)
-            auto fontURL = URL::fileURLWithFileSystemPath(String(fontsPath.get()));
+        RetainPtr fontsPath = [path stringByAppendingPathComponent:font];
+        auto fontURL = URL::fileURLWithFileSystemPath(String(fontsPath.get()));
+        NSURL *nsFontURL = fontURL;
```
I think that you can replace this with `NSURL *nsFontURL = [NSURL fileURLWithPath:[path stringByAppendingPathComponent:font]]`.

pvollan committed:
User installed fonts are not always registered
https://bugs.webkit.org/show_bug.cgi?id=286511
rdar://143598462

Reviewed by Alexey Proskuryakov.

User installed fonts are not always registered because we are doing a case sensitive comparison.

* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::registerUserInstalledFonts):

Canonical link: https://commits.webkit.org/289691@main
